src/libfetchers/git.cc

```diff
@@ -529,7 +529,13 @@ struct GitInputScheme : InputScheme
             if (input.getRev()) {
                 try {
                     runProgram("git", true, { "-C", repoDir, "--git-dir", gitDir, "cat-file", "-e", input.getRev()->gitRev() });
-                    doFetch = false;
+                    auto res = runProgram("git", true, { "-C", repoDir, "--git-dir", gitDir, "for-each-ref", fmt("--contains=%s", input.getRev()->gitRev()) });
```

Comment:
I don't know how clever `git` is about this combination of flags. A naive implementation of `for-each`, `--contains` would be slow when many refs need to be checked.

Reply:
I tested this on a repo with 45k refs on a branch with 377k commits and it took ~300ms. So, I thought I was good. But, your comment made me double-check against the `linux` kernel. That took 10m....... Something doesn't scale. I'll revert. It's not a huge deal if we don't do this, since we either have the rev (in which case we fetch the rev explicitly, like below) or we don't (in which case `doFetch = true` anyway).